File: frontend/src/app/core/services/tarea.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { TareaRequest, TareaResponse, AlumnoTareaItem } from '../models/tarea.model';

@Injectable({
  providedIn: 'root'
})
export class TareaService {
  private http = inject(HttpClient);
  private apiUrl = `${environment.apiUrl}/tareas`;

  listarPorModulo(moduloId: number): Observable<TareaResponse[]> {
    return this.http.get<TareaResponse[]>(`${this.apiUrl}/modulo/${moduloId}`);
  }

  listarTareasPorCursoParaAlumno(cursoId: number): Observable<AlumnoTareaItem[]> {
    return this.http.get<AlumnoTareaItem[]>(`${this.apiUrl}/curso/${cursoId}/alumno`);
  }

  listarTodasMisTareas(): Observable<AlumnoTareaItem[]> {
    return this.http.get<AlumnoTareaItem[]>(`${this.apiUrl}/alumno/todas`);
  }

  obtenerPorId(id: number): Observable<TareaResponse> {
    return this.http.get<TareaResponse>(`${this.apiUrl}/${id}`);
  }

  crear(request: TareaRequest): Observable<TareaResponse> {
    return this.http.post<TareaResponse>(this.apiUrl, request);
  }

  actualizar(id: number, request: TareaRequest): Observable<TareaResponse> {
    return this.http.put<TareaResponse>(`${this.apiUrl}/${id}`, request);
  }

  cambiarEstado(id: number, estado: boolean): Observable<void> {
    return this.http.patch<void>(`${this.apiUrl}/${id}/estado?estado=${estado}`, {});
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}

File: backend/src/main/java/com/insteip/backend/service/interfaces/TareaService.java
package com.insteip.backend.service.interfaces;

import com.insteip.backend.domain.dto.tarea.AlumnoTareaItemDTO;
import com.insteip.backend.domain.dto.tarea.TareaRequestDTO;
import com.insteip.backend.domain.dto.tarea.TareaResponseDTO;

import java.util.List;

public interface TareaService {

    List<TareaResponseDTO> listarPorModulo(Long moduloId);

    List<AlumnoTareaItemDTO> listarTareasPorCursoParaAlumno(Long cursoId, String correo);

    List<AlumnoTareaItemDTO> listarTodasLasTareasParaAlumno(String correo);

    TareaResponseDTO obtenerPorId(Long id);

    TareaResponseDTO crear(TareaRequestDTO request);

    TareaResponseDTO actualizar(Long id, TareaRequestDTO request);

    void cambiarEstado(Long id, Boolean estado);

    void eliminar(Long id);
}
